### ISSUE TITLE
Support for splitting single-file FLAC files with cue.

### ISSUE THREAD

**Issue by sdfg2** [Unpackerr/unpackerr#141]
Hello,

I love this, it's amazing, I was wondering if at some point it might support extracting single-file flacs using something like https://github.com/ftrvxmtrx/split2flac ?

**Comment by davidnewhall** [comment]
Unfortunately, no, that is not something this application is going to do. This app has no external dependencies, and I don't know the first thing about cutting up music files like that. Future versions will probably have triggers, so you can execute a script (like the one you linked) after an extraction is finished.

EDIT: 1/3/22: While unpackerr still doesn't support this feature, command hooks have been added. Unfortunately these files are not often compressed so the command hook never fires for these. The script below may be useful to others that stumble into this thread.

EDIT: 2/23/26: We do this now.

**Comment by sdfg2** [comment]
Ok thanks!

**Comment by mprachar** [comment]
Hello @sdfg2 Unpackerr is working great for me as well, and I actually came over here to request the same thing - did you ever find a solution?

**Comment by davidnewhall** [comment]
Do these Flac files come in a compressed archive? (rar/zip files)

**Comment by mprachar** [comment]
No, what happens is the FLAC file is one continuous file, and there will also be a second file (*.cue) that contains the instructions on how to split it into separate tracks.

There are a variety of open source libraries and command line programs that simplify the process.

**Comment by davidnewhall** [comment]
It sounds like the downloader application needs to trigger the script in this case. I still just don't know how this process would fit into what unpackerr does. Surely someone has solved this, somewhere.... Wishing you both the best of luck!

**Comment by mprachar** [comment]
Thanks for considering!

I think it seems sensible as it’s a common post-download step that needs to be done before a *arr application can complete the import process. I also understand it’s not quite a straightforward as decompression/extraction.

I’ll keep looking, meanwhile if you do wind up adding a step for your application to trigger a script I’ll give that a shot.

Best Regards,
Mike

**Comment by davidnewhall** [comment]
> I’ll keep looking, meanwhile if you do wind up adding a step for your application to trigger a script I’ll give that a shot.

I did add this, but if the Flac files are not compressed to begin with, Unpackerr won't do anything with them. I'll keep thinking about this, and more thoughts are welcomed.

**Comment by sdfg2** [comment]
> Hello @sdfg2 Unpackerr is working great for me as well, and I actually came over here to request the same thing - did you ever find a solution?

Yup!  I created a bash script that runs on completion of a torrent in transmission.

```
#!/bin/bash
#
#
#  Simple transmission script for splitting single file albums into multiple tracks.
#
#  Determines if the completed torrent is a directory, if so iterate through all directories in the torrent directory
#  looking for single flac/ape/m4a/wv/wav files.  If there's a single file, check for a matching name .cue file.
#  If there is, extract the single file album using https://github.com/ftrvxmtrx/split2flac
#

shopt -s nullglob

startdir="$TR_TORRENT_DIR/$TR_TORRENT_NAME"
#echo `date` "Download complete: $startdir"

if [ -d "$startdir" ]; then
    while IFS= read -r -d '' dir; do
        cd "$dir"
        echo "Checking $dir"
        cuefiles=`find "$dir" -type f -regextype posix-extended -regex '.*.(cue)' -printf x | wc -c`
        if [ "$cuefiles" -gt "0" ]; then
            echo "    ...$cuefiles cue files found"
            while IFS= read -r -d '' flacname; do
                flacclean=${flacname%.*}
                echo "    ...trying $flacclean"
                if [ -f "$flacclean.cue" ] || [ -f "$flacname.cue" ]; then
                    if [ -f "$flacclean.cue" ]; then
                        cueformat=`file -b --mime-encoding "$flacclean.cue"`
                    fi
                    if [ -f "$flacname.cue" ]; then
                        cueformat=`file -b --mime-encoding "$flacname.cue"`
                    fi
                    echo "    ...cueformat $cueformat"
                    echo "    ...extracting $flacname"
                    split2flac -nc -nC -d -D -F -cuecharset $cueformat "$flacname" && echo "    ...complete" || echo "    ...error"
                    find "$dir" -type f -exec chmod 664 {} \;
                    find "$dir" -type d -exec chmod 775 {} \;
                    if [ -f "00.flac" ]; then
                        rm 00.flac
                    fi
                fi
            done < <(find "$dir" -type f -regextype posix-extended -regex '.*.(flac|ape|m4a|wv|wav)' -print0)
        else
            echo "    ...nothing to do"
        fi
    done < <(find "$startdir" -type d -print0)
else
    echo "    ...not a directory, nothing to do"
fi
echo "    Done!"
```

It relies on [split2flac](https://github.com/ftrvxmtrx/split2flac).  It's a nasty bodge, and it could probably be refined/optimized/whatever up the wazoo, but this works for me.  There will be outlier cases that it fails on and requires manual intervention.  Also I don't know what's up with the encoding on some of these cue files but it will actively create multi-gigabyte files if it trips up on something - I keep meaning to put some kind of timeout and cleanup but never get round to it.

**Comment by davidnewhall** [comment]
fwiw, unpackerr has command hooks now, but it will only work on files that are compressed. That's probably useless for this feature request. The above script looks like a good place to start. I might recommend to @sdfg2 to commit that into a git repo or at least a [gist](https://gist.github.com/). This way others can contribute code and ideas. It can slowly grow into something more awesome.

I'm going to continue to leave this issue open to increase visibility. I'm open to ideas too.

**Comment by mann1x** [comment]
@davidnewhall 

From the same author of split2flac

https://github.com/nilzeronull/unflac
https://sr.ht/~ft/unflac/

Works very well, output directory can be set with arg -o
Only need to specify the .cue as input
Does have a minor issue with processing any "uncommon" character in the cue filename (including dash) so it must be sanitized

**Comment by davidnewhall** [comment]
Not super fond of calling out to `ffprobe`.

**Comment by mann1x** [comment]
Unfortunately there's nothing else that does it... at least I didn't find it

**Comment by davidnewhall** [comment]
I linked a go flac library above. Just needs time, and I don't really do audio stuff, so this is new to me.
https://github.com/mewkiz/flac/blob/v1.0.8/meta/cuesheet.go#L15

**Comment by mann1x** [comment]
I'm not really good at go so I can't help much...
Yes the library looks very complete.
But maybe, if it's appropriate, the unflac script could be something effortless to start with the aim to replace it later with something proper.

**Comment by davidnewhall** [comment]
I'm not one to do things half way unfortunately. If someone else wanted to try to make unpackerr call out to `ff` tools, I could consider merging it, but I wont spend time on that. I may spend time on the flac library at some point.

**Comment by mann1x** [comment]
Understandable.
If you give me a hint where to start, I can try.
Don't have much time as I have a few years backlog on my open source tools to maintain...
But I didn't realize how incredibly often music releases are in cue/flac.
I have to spend quite some time doing it manually and it's painful, I could invest that time in it.

**Comment by davidnewhall** [comment]
There's a pretty tight integration with the xtractr library. Unpackerr makes a call to an xtractr method to find "Extractable" files. So the first thing you'd need to do is change how that call works so it can also find flac+cue. I don't think that can go in the xtractr library, so a new function needs to be written to check for cue/flac. Once you find them, pass them into the code you linked. And by that, I mean rewrite that code into a sub-package in unpackerr. The other problem is that code has no license, so you can't really copy/paste it.

**Comment by mann1x** [comment]
Thanks, that's a good start.

The code has been published at least once with the MIT license as a package:

https://voidlinux.pkgs.org/current/voidlinux-main-aarch64/unflac-1.0_2.aarch64.xbps.html

I will ask the author to provide one.

**Comment by mprachar** [comment]
Lookig forward to this if you get it working!

Best Regards,
Mike

________________________________

**Comment by CHODEUS** [comment]
Did this ever take off? It is a well needed integration but finding the right place to integrate this tool is important.

**Comment by davidnewhall** [comment]
Nope. I don't download music, so not really something I'm looking to do here. I'm happy to brainstorm with anyone else that writes Go and wants to contribute.

**Comment by davidnewhall** [comment]
@oceanplexian added support for splitting FLAC/CUE files. Would anyone care to test this?

Install the unstable version. Instructions: https://unstable.golift.io/unpackerr
- Set split_flac to true for Lidarr. ie. `UN_LIDARR_0_SPLIT_FLAC=true`
```
[[lidarr]]
 split_flac = true
```

**Comment by chodeus** [comment]
I'll try have a look at it tomorrow!

**Comment by mprachar** [comment]
Awesome – I’ll give it a shot!
Best Regards, Mike

**Comment by JohnGalt1717** [comment]
Testing this. IT appears to do what it should and extracts them.

The issue is that lidarr will then require intervention for all imports because the original .flac is still there and as a result it thinks it's a track.

Suggest either just deleting the source entirely or renaming it to .bak at the end or something so that Lidarr doesn't get confused.

**Comment by davidnewhall** [comment]
We can fix this. @oceanplexian You wanna take a stab at it? EDIT: This may actually be more challenging than just deleting the file. We probably need to rename the file to .bak and then rename it back to .flac after Lidarr imports the content. That way a torrent client can continue to seed the data. The rename may break it, but most of them are resilient enough to either keep the file handle open, or re-open it after the file re-appears. I'll ask the Lidarr devs what their thoughts are on for this.

EDIT2: The direction I'm going to take this is: After split, call Lidarr API to import split files.

**Comment by chodeus** [comment]
Was just coming in to say that, if you delete the original flac you'll break the seed which is what seeders don't want. 
Also, good work getting this feature implemented!!

**Comment by JohnGalt1717** [comment]
Also doesn't seem to be doing APE + cue.

**Comment by davidnewhall** [comment]
No one asked of that until 10 seconds ago I've never heard of it. Do you get an error in the log saying the cue file has no supported audio format? Would you mind attaching the cue file for an APE file?

**Comment by JohnGalt1717** [comment]
With APE+Cue it doesn't seem to try. It just skips them.

This would generally need ffmpeg to convert the ape to flac and then use unflac or whatever you're doing to split it, but they're everywhere and super annoying so if you can support, it would be fantastic.

As for the deletion/rename, perhaps the outputs can go to a sub folder like unflac does by default, and then tell lidarr that the path changed?

**Comment by davidnewhall** [comment]
Yeah, APE is probably going to be a while. I don't use ffmpeg, Unpackerr doesn't call other binaries. If I can find a technical spec for the APE format, I can likely make it happen. 

I added a change so Unpackerr will attempt to import the files into Lidarr. I have done zero testing, so unstable is truly unstable at the moment. Please try it out and share your findings (log file?).

**Comment by JohnGalt1717** [comment]
Looks like ape is an actual format.... if it could just split on the .cue and keep it in the same format that would be a win.

**Comment by davidnewhall** [comment]
I do not currently have a way to make APE work. I looked into it, and I have been unable to locate a solid spec sheet on the format. There are no libraries to do any of this, so I will likely have to develop my own. That'll take some time. 

Have you been able to test if FLAC imports now?

**Comment by JohnGalt1717** [comment]
Gotta get one to see it work. I'll keep you posted.

**Comment by JohnGalt1717** [comment]
[INFO] 2026/02/24 09:39:42 [Radarr] Updated (http://localhost:7878): 0 Items Queued, 0 Retrieved
[INFO] 2026/02/24 09:39:42 [Lidarr] Updated (http://localhost:8686): 1 Items Queued, 1 Retrieved
[INFO] 2026/02/24 09:39:42 [Sonarr] Updated (http://localhost:8989): 0 Items Queued, 0 Retrieved
[INFO] 2026/02/24 09:40:28 [Unpackerr] Queue: 0 waiting, 0 queued, 0 extracting, 0 extracted, 0 imported, 0 failed, 0 deleted
[INFO] 2026/02/24 09:40:28 [Unpackerr] Totals: 13 retries, 11 finished, 0|0 webhooks, 0|0 cmdhooks, stacks; event:0, hook:0, del:0, up 4 days 22 hours 42 mins

I think this means that Lidarr queued one. Not sure how to verify anything more.

**Comment by davidnewhall** [comment]
13 retries, hm. Please upload the log file.

**Comment by JohnGalt1717** [comment]
[unpackerr.log](https://github.com/user-attachments/files/25528042/unpackerr.log)

You go. I sanitized pathing etc, but otherwise untouched.

**Comment by davidnewhall** [comment]
This is only 11 minutes and doesn't contain any of the 13 failed extractions. Check older files? You can jump on Discord and DM me a log if that's easier.

**Comment by kontell** [comment]
I've tested v0.15 and I've found that completed downloads in Lidarr's queue are not processed.

When I move a download to my watch folder unpackerrr does split the flac. However, I've got some playback issues. VLC plays the split flac files but the progress bar is messed up. The Strawberry media player doesn't play the files at all and gives the below error.

I checked the original flac with shntool, the files are split and played as expected.

> Internal data stream error.
../libs/gst/base/gstbaseparse.c(3754): gst_base_parse_loop (): /GstPlayBin3:pipeline-1-pipeline/GstURIDecodeBin3:uridecodebin3/GstURISourceBin:urisourcebin0/GstParseBin:parsebin0/GstFlacParse:flacparse0: streaming stopped, reason error (-5)

[unpackerr.log](https://github.com/user-attachments/files/25717453/unpackerr.log)

[unpackerr.conf.txt](https://github.com/user-attachments/files/25717461/unpackerr.conf.txt)

**Comment by davidnewhall** [comment]
> I've found that completed downloads in Lidarr's queue are not processed.

This is because someone change the protocol from "torrent" and "usenet" to some other string. You need to set protocols variable correctly for Lidarr. I'll get the docs updated soon. There's an open issue on the unpackerr repo for this.

>  media player doesn't play 

To be honest, we've never tested this code. I cannot find a flac file to split. I'll keep looking because this is the feedback I need and the bugs I want to fix. Thank you!

**Comment by davidnewhall** [comment]
@kontell We've loaded some fixes into the unstable branch. Can you try again please? Lidarr should pick up torrents again by default, and hopefully the split flac files are still in good shape. Directions here: https://unstable.golift.io/unpackerr

fwiw, I'm also still trying to 'fix' it so it makes Lidarr import the split files. The code is there, but it's probably not quite working yet, so please test that too as you can.

**Comment by kontell** [comment]
You guys work quick, eh! I had a quick test with the unstable version (Version: 0.15.0-930).

The split flac files play without issue now.

The album didn't import into Lidarr:

> [ERROR] 2026/03/04 00:01:03.775770 lidarr.go:186: [Lidarr] Manual import list failed for /media/bluecon/download/lidarr/Orff - Carmina Burana (Previn): api.Get(v1/manualimport): decoding Starr JSON response body: json: cannot unmarshal string into Go struct field AudioTags.audioTags.duration of type starr.TimeSpan

I'll have another look in the morning and get some logs from Lidarr if available.

[unpackerr_unstable.log.txt](https://github.com/user-attachments/files/25729282/unpackerr_unstable.log.txt)

**Comment by davidnewhall** [comment]
This is great feedback. I'll have another push for you to test shortly. Don't wait up for me though if you're headed off soon.

**Comment by davidnewhall** [comment]
Pushed a [fix](https://github.com/golift/starr/pull/188/changes/8fd76a6486047bff3ff9792b9224e47531eea90e) into unstable. It's ready to test now.

**Comment by kontell** [comment]
Another quick test, an attempt at import is triggered this time. But it appears to be trying to import the 25 split files plus the original flac.

There can be numerous other reasons why Lidarr would reject an import. Track tag/ naming being off and slight variations between album releases (versions). So this wouldn't have imported even if only the 25 split files are sent to Lidarr but that's obviously not your issue.

I'll have a play with some other albums later today.

[manual_import_lidarr.txt](https://github.com/user-attachments/files/25736856/manual_import_lidarr.txt)

[unpackerr.2026-03-04.log.txt](https://github.com/user-attachments/files/25736857/unpackerr.2026-03-04.log.txt)

**Comment by davidnewhall** [comment]
Thank you for the logs! This looks promising. I'll see if I can make it not try to import the original file too. And I fully understand that some things are simply not going to import. I'm just doing what I can since Lidarr won't. Will examine this new data soon!

**Comment by kontell** [comment]
It'd be good if the unpackerr triggered import showed up in the activity history when it fails, currently I just see the first attempt from Lidarr when the file is downloaded.

**Comment by davidnewhall** [comment]
I'm not sure if there's anything I can do to make a failed manual import show up in the history. I asked in the Servarr discord 🤞 

I pushed [another update](https://github.com/Unpackerr/unpackerr/commit/cead14643dad6e3400fcbfefc8382d5b66ad2ead) to unstable. It will now only attempt to import files that it actually split.

**Comment by kontell** [comment]
It's mostly working now and I've had a few albums import just fine.

> I'm not sure if there's anything I can do to make a failed manual import show up in the history

Thought that might be awkward as "manual" import showing up in the activity log wouldn't be natural.

There's some quirkiness causing issues. The cue sheet often references a .wav file instead of .flac. 

> [ERROR] 2026/03/05 13:43:44.441962 handlers.go:230: Extraction Failed: foobar: extraction failed: audio file referenced by cue sheet not found: /media/bluecon/download/lidarr/foobar/_CDImage.wav; unknown archive file type: /media/bluecon/download/lidarr/foobar/CDImage.cue (file: /media/bluecon/download/lidarr/foobar/CDImage.cue)

According to a chatbot this is quite common (shntool handles it without issue).

> Yes — that’s completely normal.
Why the .cue references .wav
Your cue was generated by Exact Audio Copy.
EAC’s traditional workflow is:
Rip the CD → single WAV image
Generate cue sheet referencing that WAV
Optionally compress the WAV → FLAC / APE / etc
So the cue sheet preserves the original WAV reference, even if the audio file was later encoded to FLAC.

I've also got some albums that are stalled at "Extracted, Awaiting Import", no idea why that started happening.

[unpacker.2026-03-05.log](https://github.com/user-attachments/files/25771760/unpacker.2026-03-05.log)

**Comment by davidnewhall** [comment]
Thank you for the follow up!. I'll see if we can add WAV support....

> I've also got some albums that are stalled at "Extracted, Awaiting Import", no idea why that started happening.

Look in the Lidarr queue. Why aren't they importing?

**Comment by kontell** [comment]
You don't need support for the WAV format, just need to look for a .flac when the .cue file refers to a .wav (to be clear, there are no WAV files in the torrents).

In the Lidarr queue I only see the reason for the original import attempt failing (trying to import the single flac file).

**Comment by davidnewhall** [comment]
So the cue says wav, but it's actually flac? Same name except for the extension?

If Unpackerr sees things in your Lidarr queue that you do not see, turn on this option. Click Options in the top right while looking at the Activity Queue.

<img width="655" height="126" alt="Image" src="https://github.com/user-attachments/assets/931fcecb-e3f5-40ba-b76b-1ee434e78ea3" />

**Comment by JohnGalt1717** [comment]
Note that this could say "wav" and be .ape as well.

In the case of .ape instead of .flac, should still be able to split it using that format and just create .ape files instead of .flac files.

**Comment by davidnewhall** [comment]
Just need a spec on APE files. Any idea where I find that?

**Comment by kontell** [comment]
> So the cue says wav, but it's actually flac? Same name except for the extension?

Exactly.

**Comment by kontell** [comment]
> I've also got some albums that are stalled at "Extracted, Awaiting Import", no idea why that started happening.

Sorry, I got confused. The manual import was triggered by unpackerr and failed for naming reasons or whatever.

The "Item Waiting for Import" later in the logs is just waiting for the user to intervene? (I thought unpackerr was waiting).

I had Show Unknown Artist Items ticked, it's just that the error shown on hover only refers to the earlier single flac attempt. 

The import is failing on most albums due to the split files not having any album tags. This can be got from the Title in the .cue sheet. Adding the artist to the track tags as well would probably help.

**Comment by davidnewhall** [comment]
> The "Item Waiting for Import" later in the logs is just waiting for the user to intervene? 

Yes, it's just informational. It can't find anything to extract (or split?) (or it already did), and it's waiting for Lidarr to import it. EDIT: Keep in mind this code used to just extract and wait for lidarr/sonarr to import. Now, with flac files, Lidarr will never try to import them, so I have built code to make Unpackerr 'try'. The log message could probably be more clear and say "manual import failed, do it yourself" instead of "waiting for import"

> This can be got from the Title in the .cue sheet. Adding the artist to the track tags as well would probably help.

Can you attach a sample cue sheet?

**Comment by davidnewhall** [comment]
Updated unstable so it should now tag the files. edit: ....and now it will look for flac if wav is missing.
edit 40 minutes later: now it extracts the album cover from the flac (if it's there), and embeds it in all the written tracks as well as writes it to disk.

**Comment by kontell** [comment]
The tagging seems fine now.

On the two albums I tried the files were extracted, seemingly successfully. But then deleted due to a file not found error without an attempt to trigger a manual import.

[unpacker.2026-03-06.log](https://github.com/user-attachments/files/25794725/unpacker.2026-03-06.log)

**Comment by davidnewhall** [comment]
This is another bug I introduced and will find a way to fix.

**Comment by kontell** [comment]
You're very close though, made me laugh at the files being created and then deleted for no apparent reason.

**Comment by davidnewhall** [comment]
Pushed a fix. Unstable building now. Ready shortly. Please test again!

**Comment by kontell** [comment]
This is working well now. I appreciate your efforts here as I know it's not easy without being able to test it.

I did come across some UTF/ file naming issues. One of them I think is Lidarr so I don't think there's much you can do other than sanitise the file names for Lidarr (see the bottom of the log).

With the other one the cue sheet is in UTF-16 so even nano on my machine couldn't open it successfully (in log).

Another album had an "O" in the cue sheet and a "Ö" in the file name causing the file not to be found.

I'm not sure how far you want to go in dealing with all these little quirks but I think it's working very well now.

[unpacker.2026-03-07.log](https://github.com/user-attachments/files/25814311/unpacker.2026-03-07.log)

**Comment by davidnewhall** [comment]
It's difficult to find people willing to do this testing and back-n-forth. Thank YOU!

- As for the accented characters in cue but not in filename... that's a bad rip and not something I can easily fix. re: Ö
- For Blue Oyster Cult, can you zip the cue file and attach that directly? I want to make sure `cat` + copy/paste didn't lose anything. I can probably fix that one.
- This is probably also something I can fix: '04 - It'$'\222''s Hard to Find a Way.flac' - will look into it. edit: maybe attach that zip'd cue file too?

**Comment by kontell** [comment]
Ya, I don't know if it's worthwhile trying ASCII folding, cue sheets attached.

[Cue-Sheets.zip](https://github.com/user-attachments/files/25815783/Cue-Sheets.zip)

**Comment by davidnewhall** [comment]
The flac file that had the accented O... Did the cue and the flac have identical names? As in, could I use the cue's name as a way to find the flac?

```
Blue Öyster Cult - Club Ninja.cue
Blue �yster Cult - Club Ninja.flac
```

Looks like it.

**Comment by kontell** [comment]
Ya, the only difference was the accent.

**Comment by kontell** [comment]
I wan't clear. From what I've seen the files are always named the same, so you should be able to find it like that.

**Comment by davidnewhall** [comment]
I think I fixed all 3 problems.  Unstable building now, ready in 10m or so. (check date on unstable website)

**Comment by kontell** [comment]
All those fixes worked.

Regarding the  "Ö" fix, there was a second cue sheet in the directory (*.fixed.cue) where the file name was corrected. So I don't think trying to allow for errors in the .cue sheet is worthwhile (it extracted the files twice, but no harm done).

Everything working for me, I don't have anything else. Thanks.

**Comment by davidnewhall** [comment]
That sounds better than not at all, or deleting them after. I'll probably just leave it like it is until it causes a real problem.

Thanks so much! I'll push a new release today or tomorrow.

**Comment by davidnewhall** [comment]
I'm going to Lock this issue so it doesn't continue to grow. There's a lot of good history in here. If you run into problems extracting CUE files, please open another issue.

EDIT: 3/8/2026 -> The fixes from this issue have [been released](https://github.com/Unpackerr/unpackerr/releases/tag/v0.15.1).
